File: package/server/app/schemas/annual_report.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel

from app.schemas.photo import Photo


class UserInfo(BaseModel):
    nickname: str
    avatarUrl: str

class TimeMetrics(BaseModel):
    totalPhotos: int
    accompanyDays: int
    firstPhotoDate: Optional[str] = None
    lastPhotoDate: Optional[str] = None
    lateNightPhotoCount: Optional[int] = 0
    photoDates: List[str] = []

class CategoryDistributionItem(BaseModel):
    name: str
    value: int

class MemoryMetrics(BaseModel):
    categoryDistribution: List[CategoryDistributionItem]
    topPersonName: str
    topPersonCount: int
    topLocation: str
    maxPhotoDay: str
    maxPhotoDayCount: int
    topFeature: str
    topFeatureCount: int
    topMake: str
    topModel: str
    topMakeModelCount: int

class CarouselGroup(BaseModel):
    id: str
    locationName: str
    photos: List[str]

class EmotionMetrics(BaseModel):
    livePhotos: int
    backupPhotos: int
    totalVideoDuration: float
    cameraPhotos: int
    emotionCarouselGroups: List[CarouselGroup]

class TopCity(BaseModel):
    cityName: str
    photoCount: int
    provinceName: str

class LocationPoint(BaseModel):
    lng: float
    lat: float
    name: str
    count: int
    coverUrl: Optional[str] = None

class LocationMetrics(BaseModel):
    lightenProvinceNum: int
    lightenCityNum: int
    topCities: List[TopCity]
    locationPoints: List[LocationPoint]
    farthestCity: Optional[str] = None
    farthestDistance: Optional[int] = 0
    farthestCityPhotos: List[Photo] = []

class SeasonData(BaseModel):
    seasonName: str
    photoCount: int
    topTag: str
    representativePhoto: str
    highlight: str
    shootMonth: str

class SeasonMetrics(BaseModel):
    seasonList: List[SeasonData]

class MonthlyExpense(BaseModel):
    month: str
    amount: float

class ExpenseMetrics(BaseModel):
    totalAmount: float
    totalCount: int
    averagePrice: float
    monthlyTrend: List[MonthlyExpense]
    totalAmountLastYear: Optional[float] = 0.0
    monthlyTrendLastYear: List[MonthlyExpense] = []
    maxExpenseTicket: Optional[str] = None
    maxExpenseAmount: Optional[float] = 0

class MonthlyFrequency(BaseModel):
    month: str
    count: int

class RouteStats(BaseModel):
    route: str
    count: int

class DestinationStats(BaseModel):
    city: str
    count: int

class TripTypeDistribution(BaseModel):
    workday: int
    weekend: int
    holiday: int

class TravelBehaviorMetrics(BaseModel):
    monthlyFrequency: List[MonthlyFrequency]
    topRoutes: List[RouteStats]
    topDestinations: List[DestinationStats]
    tripTypeDistribution: TripTypeDistribution

class ComprehensiveMetrics(BaseModel):
    totalMileage: int
    costPerKm: float

class TransportAnalysisMetrics(BaseModel):
    behavior: TravelBehaviorMetrics
    comprehensive: ComprehensiveMetrics

class TicketDetail(BaseModel):
    id: str
    train_code: str
    departure_station: str
    arrival_station: str
    date_time: datetime
    price: float
    seat_type: str
    name: str

class EasterEggTags(BaseModel):
    main: str
    sub: List[str]

class EasterEgg(BaseModel):
    bestPhotoUrl: str
    bestPhotoDate: str
    tags: EasterEggTags

class AnnualReportData(BaseModel):
    year: int
    user: UserInfo
    time: TimeMetrics
    memory: MemoryMetrics
    emotion: EmotionMetrics
    location: LocationMetrics
    season: SeasonMetrics
    easterEgg: EasterEgg
    expense: Optional[ExpenseMetrics] = None
    travelBehavior: Optional[TravelBehaviorMetrics] = None
    comprehensive: Optional[ComprehensiveMetrics] = None
    transportAnalysis: Optional[TransportAnalysisMetrics] = None

File: package/website/src/types/annualReport.ts
import { Photo } from "./album";

export interface UserInfo {
  nickname: string;
  avatarUrl: string;
}

export interface TimeMetrics {
  totalPhotos: number;
  accompanyDays: number;
  firstPhotoDate?: string;
  lastPhotoDate?: string;
  lateNightPhotoCount?: number;
  photoDates: string[];
}

export interface MemoryMetrics {
  categoryDistribution: { name: string; value: number }[];
  topPersonName: string;
  topPersonCount: number;
  topLocation: string;
  maxPhotoDay: string; // Date string
  maxPhotoDayCount: number;
  topFeature: string;
  topFeatureCount: number;
  topMake: string;
  topModel: string;
  topMakeModelCount: number;
}

export interface CarouselGroup {
  id: string;
  locationName: string;
  photos: string[];
}

export interface EmotionMetrics {
  starredPhotos: number;
  backupPhotos: number;
  totalOpenTimes: number;
  totalVideoDuration: number; // 总视频播放时长（秒）
  starredPhotosList: string[]; // URLs
  sharedPhotosList: string[]; // URLs
  emotionCarouselGroups: CarouselGroup[]; // New field for carousel
}

export interface LocationMetrics {
  lightenProvinceNum: number; // 年度点亮省份数量（核心）
  lightenCityNum: number;      // 年度点亮地级市数量（核心）
  topCities: Array<{           // 年度打卡TOP3城市（含照片数）
    cityName: string;
    photoCount: number;
    provinceName: string;
  }>;
  locationPoints: Array<{      // 地域点位列表（地图可视化用，含经纬度）
    lng: number;               // 经度
    lat: number;               // 纬度
    name: string;              // 地点名称
    count: number;             // 该地点拍摄照片数
    coverUrl?: string;         // 城市封面图
  }>;
  farthestCity: string; // 年度最远打卡城市
  farthestDistance: number; // 年度最远打卡城市距离（公里）
  farthestCityPhotos: Photo[]; // 年度最远打卡城市照片URL列表
}

export interface SeasonData {
  seasonName: '春' | '夏' | '秋' | '冬'; // 季节名称
  photoCount: number; // 该季节照片数量
  topTag: string; // 该季节TOP1智能标签（如春→“嫩芽”、夏→“蝉鸣”）
  representativePhoto: string; // 该季节代表性照片URL（取首图/高赞图）
  highlight: string; // 该季节核心亮点（人物/场景，如“和好友的12次夏日露营”）
  shootMonth: string; // 该季节高频拍摄月份（如“3-5月”）
}

export interface SeasonMetrics {
  seasonList: SeasonData[]; // 春/夏/秋/冬四季节数据数组
}

export interface EasterEgg {
  bestPhotoUrl: string;
  bestPhotoDate: string;
  tags: {
    main: string;
    sub: string[];
  };
}

export interface MonthlyExpense {
  month: string;
  amount: number;
}

export interface ExpenseMetrics {
  totalAmount: number;
  totalCount: number;
  averagePrice: number;
  monthlyTrend: MonthlyExpense[];
  totalAmountLastYear?: number;
  monthlyTrendLastYear?: MonthlyExpense[];
  maxExpenseTicket?: string;
  maxExpenseAmount?: number;
}

export interface MonthlyFrequency {
  month: string;
  count: number;
}

export interface RouteStats {
  route: string;
  count: number;
}

export interface DestinationStats {
  city: string;
  count: number;
}

export interface TripTypeDistribution {
  workday: number;
  weekend: number;
  holiday: number;
}

export interface TravelBehaviorMetrics {
  monthlyFrequency: MonthlyFrequency[];
  topRoutes: RouteStats[];
  topDestinations: DestinationStats[];
  tripTypeDistribution: TripTypeDistribution;
}

export interface ComprehensiveMetrics {
  totalMileage: number;
  costPerKm: number;
}

export interface TransportAnalysisMetrics {
  behavior: TravelBehaviorMetrics;
  comprehensive: ComprehensiveMetrics;
}

export interface TicketDetail {
  id: string;
  train_code: string;
  departure_station: string;
  arrival_station: string;
  date_time: string;
  price: number;
  seat_type: string;
  name: string;
}

export interface AnnualReportData {
  year: number;
  user: UserInfo;
  time: TimeMetrics;
  memory: MemoryMetrics;
  emotion: EmotionMetrics;
  location: LocationMetrics;
  season: SeasonMetrics;
  easterEgg: EasterEgg;
  expense?: ExpenseMetrics;
  travelBehavior?: TravelBehaviorMetrics;
  comprehensive?: ComprehensiveMetrics;
  transportAnalysis?: TransportAnalysisMetrics;
}
